src/codegen.c

```diff
@@ -918,29 +909,28 @@ void codegen(Program *prog, bool do_opt) {
       if (!fn->params->next || (fn->params->next && fn->params->next->next)) {
         error("main() has incorrect parameter count: 0 or 2 expected");
       }
-      if (fn->params->var->ty->kind != TY_INT ||
-          fn->params->next->var->ty->kind != TY_PTR || // argv[][] == **argv
-          !fn->params->next->var->ty->base ||
-          fn->params->next->var->ty->base->kind != TY_PTR ||
-          fn->params->next->var->ty->base->base->kind != TY_CHAR) {
+      Type* ty1 = fn->params->var->ty;
+      Type* ty2 = fn->params->next->var->ty;
+      if (ty1->kind != TY_INT || ty2->kind != TY_PTR || !ty2->base ||
+          ty2->base->kind != TY_PTR || ty2->base->base->kind != TY_CHAR) {
         error("main() parameters have incorrect types: main(int argc, char *argv[]) expected");
       }
       need_argc_argv = true;
     }
   }
   if (!need_argc_argv) {
     printf("  LIT2r 0000 main_ POP2r BRK\n");
+  } else {
     if (need_device_hook[0]) {
```

Comment:
I'm not sure how you'd handle this, but this is also varvara-specific isn't it? (Sorry, I knew the `[0]` was a messy solution when I wrote this.)

Reply:
Yeah, the whole argc/argv hook is specific to varvara (even if only the console read port would need to be changed between implementations... or it may be more than that, because I didn't grok the hook code yet). Why did the check exist in the first place? The TODO seems to confirm that cli arguments and a console hook don't clash fundamentally, so is this just a temporary limitation?

In the meantime, I've made the function to generate argc/argv comprehension specific to varvara and an argument to codegen. I tried to resolve the fixed index by letting the caller optionally specify an element of the device array as a console-type device that would clash with argc/argv.

Reply:
> Why did the check exist in the first place? The TODO seems to confirm that cli arguments and a console hook don't clash fundamentally, so is this just a temporary limitation?

Yes, it's just unfinished. The current implementation has no passthrough for calling on_console, so I want to warn the user if they try to use it.

Thanks for the cleanup!